### PR TITLE
mgr/dashboard: add debug mode

### PR DESCRIPTION
**Pull request by epuertat**
New command `dashboard debug <status|enable|disable>` added. By default, debug is disabled. This now hides Python tracebacks in HTTP responses (which could reveal sensitive details of the underlying environment, like python version, filesystem locations, etc.):

```diff
{"status": "500 Internal Server Error",
"detail": "The server encountered an unexpected condition which prevented it from fulfilling the request.",
- "traceback": "Traceback (most recent call last):
-  File "/usr/lib/python3.7/site-packages/cherrypy/_cprequest.py", line 670, in respond response.body = self.handler()
-  File "/usr/lib/python3.7/site-packages/cherrypy/lib/encoding.py", line 220, in __call__↵    self.body = self.oldhandler(*args, **kwargs)↵
-  File "/usr/lib/python3.7/site-packages/cherrypy/_cptools.py", line 237, in wrap↵    return self.newhandler(innerfunc, *args, **kwargs)↵
-  File "/ceph/src/pybind/mgr/dashboard/services/exception.py", line 88, in dashboard_exception_handler↵    return handler(*args, **kwargs)↵
-  File "/usr/lib/python3.7/site-packages/cherrypy/_cpdispatch.py", line 60, in __call__↵    return self.callable(*self.args, **self.kwargs)↵  
-  File "/ceph/src/pybind/mgr/dashboard/controllers/__init__.py", line 645, in inner↵    ret = func(*args, **kwargs)↵ 
-  File "/ceph/src/pybind/mgr/dashboard/controllers/__init__.py", line 838, in wrapper↵    return func(*vpath, **params)↵  
-  File "/ceph/src/pybind/mgr/dashboard/controllers/mgr_modules.py", line 24, in list↵    always_on_modules = mgr_map['always_on_modules'][mgr.release_name]↵  
-  File "/ceph/src/pybind/mgr/dashboard/__init__.py", line 33, in __getattr__↵    return getattr(self._mgr, item)↵  
-  File "/ceph/src/pybind/mgr/mgr_module.py", line 594, in release_name↵    return self._ceph_get_release_name()↵AttributeError: 'Module' object has no attribute '_ceph_get_release_name'↵"
+ "traceback": "",
"version": "8.9.1"}   
```
Fixes: https://tracker.ceph.com/issues/41990
Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

**Comment by epuertat** [comment]
jenkins render docs

**Comment by bk201** [comment]
Tested and works for me.

The PR doesn't fix the statement in the description `..., but also pollutes logs with traceback lines, ...` of issue. But I do like to keep the traceback in backend when there is an uncaught exception.

To ease debugging when `debug` is disabled, maybe we can also:
- Return the [request ID](https://github.com/cherrypy/cherrypy/blob/2fc78dc9af3e556829ce69d3e1311e5629200882/cherrypy/_cprequest.py#L504) of the request that raises the Exception in response
- Logging the request IDs in the backend

In this way, an end-user can still report the uuid and it's much easier to find the corresponding error in the log. Although I'm not sure how much performance impact will occur when turning the request ID on, it's lazy-loaded now.

**Comment by votdev** [comment]
The PR should make sure that debug is enabled in Teuthology and our test infra.

**Comment by rjfd** [comment]
@epuertat why is this feature implemented as a plugin?

**Comment by epuertat** [comment]
> Tested and works for me.

Thanks a lot for the thorough review, @bk201!

> The PR doesn't fix the statement in the description `..., but also pollutes logs with traceback lines, ...` of issue. But I do like to keep the traceback in backend when there is an uncaught exception.

Ooops, I assumed the tracebacks also dissapeared from the logs :(

> To ease debugging when `debug` is disabled, maybe we can also:
> * Return the [request ID](https://github.com/cherrypy/cherrypy/blob/2fc78dc9af3e556829ce69d3e1311e5629200882/cherrypy/_cprequest.py#L504) of the request that raises the Exception in response
> * Logging the request IDs in the backend

Like the idea, an uuid would work, yeah. Nevertheless, I think we should catch & process as many as exceptions we expect to happen, and return user-friendly (even non-dashboard developer friendly) messages, and leave only raw tracebacks for the unexpected.
 
> In this way, an end-user can still report the uuid and it's much easier to find the corresponding error in the log. Although I'm not sure how much performance impact will occur when turning the request ID on, it's lazy-loaded now.

Lazy-loaded? Don't get that.

**Comment by epuertat** [comment]
> The PR should make sure that debug is enabled in Teuthology and our test infra.

Sure. I'll check that. Thanks!

**Comment by epuertat** [comment]
> @epuertat why is this feature implemented as a plugin?

@rjfd: I analyzed the different alternatives:
- **In-place changes**: `module.py` is already 500 lines-of-code long.
- **External module**: it still involves adding `Options` to `module.py` + `CLICommands` + patching the config + adding an asynchronous callback. So the same stuff the plugins provide.
- **Plugins**: are best suited for extending program behaviours with minimal coupling. The plugin infrastructure was already there, so it required minimal changes to Dashboard core code, and can be easily unplugged to recover the legacy behaviour.

If you miss any other alternative, please don't hesitate to suggest. I'm sure the plugin implementation can/should be heavily improved (still requires lots of boilerplate/tricky code), so if you want to review that part too, I'm happy with amending it as well.

**Comment by bk201** [comment]
> > In this way, an end-user can still report the uuid and it's much easier to find the corresponding error in the log. Although I'm not sure how much performance impact will occur when turning the request ID on, it's lazy-loaded now.
> 
> Lazy-loaded? Don't get that.

Sorry about the wording, I mean the `uuid` is [generated only when it is accessed](https://github.com/cherrypy/cherrypy/blob/2fc78dc9af3e556829ce69d3e1311e5629200882/cherrypy/_cprequest.py#L905-L922). Not sure about the overhead of `uuid` computation. There should be no impact if we only use it in exceptions.

BTW, OpenStack services [return request ID](https://docs.openstack.org/api-guide/compute/faults.html) in the response header for every request. The request ID is quite useful when mapping requests to backend logs.

**Comment by epuertat** [comment]
> Sorry about the wording, I mean the `uuid` is [generated only when it is accessed](https://github.com/cherrypy/cherrypy/blob/2fc78dc9af3e556829ce69d3e1311e5629200882/cherrypy/_cprequest.py#L905-L922). Not sure about the overhead of `uuid` computation. There should be no impact if we only use it in exceptions.

Ok, gotcha @bk201 ! I didn't know the uuid was already there! So no big changes required then. I think the overhead of generating a uuid compared to the whole HTTPs + Ceph-mgr stack is negligible.

**Comment by rjfd** [comment]
> If you miss any other alternative, please don't hesitate to suggest. I'm sure the plugin implementation can/should be heavily improved (still requires lots of boilerplate/tricky code), so if you want to review that part too, I'm happy with amending it as well.

I would favor the "In-place changes" approach. The code would be simpler and easier to review.

Now a completely different thing (not related with the implementation approach), does this implementation also disables the traceback of exceptions raised by the controllers? (like when a `DashboardException` is raised)

**Comment by epuertat** [comment]
> I would favor the "In-place changes" approach. The code would be simpler and easier to review.

If it's not easy to review, that's my bad then. I should clean-up a bit the plugin infra, and improve docs I guess.

Regarding the simplicity: I don't think it'd be much more simple (to do exactly the same). The plugin contains the minimal code changes required in a single place (by doing it in-line, it'd require changes in multiple places, making the change harder to merge/back-port/maintain):
- **Modified behaviour**: patch `cherrypy.config` during initialization and on user-driven changes.
- **Persistency**: ceph config option `mgr/dashboard/debug`.
- **CLI User Interface**: `dashboard debug <enable|disable|status>`.

Take into account that plug-ins bring other benefits. They're easier to:
- Develop for teams, as chances of code conflicts and dependencies are reduced.
- Review and maintain/backport: all changes are in a single file.
- Easier to test and debug: new code can be easily disabled and the scope of changes is better delimited.

> Now a completely different thing (not related with the implementation approach), does this implementation also disables the traceback of exceptions raised by the controllers? (like when a `DashboardException` is raised)

Nope, this only hides the traceback details in the HTTP responses.

**Comment by sebastian-philipp** [comment]
> > If you miss any other alternative, please don't hesitate to suggest. I'm sure the plugin implementation can/should be heavily improved (still requires lots of boilerplate/tricky code), so if you want to review that part too, I'm happy with amending it as well.
> 
> I would favor the "In-place changes" approach. The code would be simpler and easier to review.

:+1:

**Comment by votdev** [comment]
When we disable Python tracebacks by default, don't you think that this will result in bugreports with useless information or makes it more harder to identify bugs? Sometimes bugs are not reproducable, so tracebacks are really helpful when they are included out-of-the-box when an error occurs. On the other hand i fully understand that the REST API should not forward sensitive data to the outer world. It is a thin line between comfort and safety.

**Comment by epuertat** [comment]
> When we disable Python tracebacks by default, don't you think that this will result in bugreports with useless information or makes it more harder to identify bugs? Sometimes bugs are not reproducable, so tracebacks are really helpful when they are included out-of-the-box when an error occurs. On the other hand i fully understand that the REST API should not forward sensitive data to the outer world. It is a thin line between comfort and safety.

I get your concern, but I don't think user (or, worse, non-user) facing HTTP responses is the place to put that information. Logs are IMHO the place to store those data. We are populating both our logs & HTTP responses with tracebacks of expected situations (like Auth failed), and I think it's extremely risky that for someone that has failed to log in to our system, we are giving them hints on the system setup, packages, versions, and filesystem locations.

**Comment by votdev** [comment]
> > When we disable Python tracebacks by default, don't you think that this will result in bugreports with useless information or makes it more harder to identify bugs? Sometimes bugs are not reproducable, so tracebacks are really helpful when they are included out-of-the-box when an error occurs. On the other hand i fully understand that the REST API should not forward sensitive data to the outer world. It is a thin line between comfort and safety.
> 
> I get your concern, but I don't think user (or, worse, non-user) facing HTTP responses is the place to put that information. Logs are IMHO the place to store those data. We are populating both our logs & HTTP responses with tracebacks of expected situations (like Auth failed), and I think it's extremely risky that for someone that has failed to log in to our system, we are giving them hints on the system setup, packages, versions, and filesystem locations.

If we are logging errors and tracebacks, then i'm fine with this :-)

**Comment by epuertat** [comment]
Comments addressed:
- `unique_id`/`uuid` is a recent feature in Cherrypy, but I implemented a fallback mechanism for older versions (less efficient, as it does not use lazy evaluation for the UUID, but given dashboard is not focused on achieving 10k request/second, it may work).
- No need of changing anything in Teuthology, as long as Traceback as still printed to logs.

File: src/pybind/mgr/dashboard/tools.py [hunk 1]
```diff
@@ -49,8 +49,7 @@ def request_begin(self):
         req = cherrypy.request
         user = JwtManager.get_username()
         # Log the request.
-        logger.debug('[%s:%s] [%s] [%s] %s', req.remote.ip, req.remote.port,
-                     req.method, user, req.path_info)
+        self._request_log(logger.debug)
```
Comment (rjfd) [inline src/pybind/mgr/dashboard/tools.py]: Logging the request is different from logging the response (the `_request_log` method name is misleading as it is actually logging the response of the "request"). In the request there is no "status", "latency", or content length. You should revert this change.

Comment (epuertat) [reply src/pybind/mgr/dashboard/tools.py]: We are printing 2 lines that repeat the same 4 fields (req.remote.ip, req.remote.port, req.method, req.path_info) and only differ in 3.

Most webservers  ([apache](https://httpd.apache.org/docs/2.4/logs.html#page-header) and [nginx](https://docs.nginx.com/nginx/admin-guide/monitoring/logging/)) only log one line per request-response pair (including request data + "status" or optionally other response data). This is an Apache access log line ([format](https://httpd.apache.org/docs/2.4/mod/mod_log_config.html#formats)):

`127.0.0.1 - peter [9/Feb/2017:10:34:12 -0700] "GET /sample-image.png HTTP/2" 200 1479`

I think we should do that and also use a separate file for event/access log with the [W3C log format](https://www.w3.org/TR/WD-logfile.html). I'll revert this and create a ticket with this proposal.

**Comment by epuertat** [comment]
> @epuertat the changes look good to me, but could you split the changes in several commits? Currently it's a single commit that changes the plugin infrastructure

Done.

**Comment by epuertat** [comment]
jenkins test docs

File: src/pybind/mgr/dashboard/HACKING.rst [hunk 2]
```diff
@@ -1844,19 +1844,40 @@ In order to create a new plugin, the following steps are required:
...
+- ``HasCommands``: requires overriding ``register_commands()`` hook by defining
+  the commands the plug-in can handle and decorating them with ``@CLICommand``.
+  The commands can be optionally returned, so that they can be invoked
+  externally (which makes unit testing easier).
+- ``HasControllers``: requires overriding ``get_controllers()`` hook by defining
+  and returning the controllers as usual.
+- ``FilterRequest.BeforeHandler``: requires overriding
+  ``filter_request_before_handler()`` hook. This method receives a
+  ``cherrypy.request`` object for processing. A usual implementation of this
+  method will allow some requests to pass or will raise a ``cherrypy.HTTPError`
```
Comment (Devp00l) [inline src/pybind/mgr/dashboard/HACKING.rst]: You missed a hyphen at the end of the line

```suggestion
  method will allow some requests to pass or will raise a ``cherrypy.HTTPError``
```

Comment (epuertat) [reply src/pybind/mgr/dashboard/HACKING.rst]: Thanks a lot for reviewing this. Good catch. How is it possible that the `rstchecker` didn't notice it. I'll look at it.

**Comment by epuertat** [comment]
jenkins render docs
